File: Part 2/csharp_solution/Exchange/Order.cs
using System;
using System.Collections.Generic;

namespace Exchange
{
    public class Order
    {
        enum Token {
            Participant,
            Instrument,
            Quantity,
            Price
            
        };

        static readonly Dictionary<int, string> _strings = new Dictionary<int, string>();

        static string GetString(ReadOnlySpan<char> span)
        {
            var hash = string.GetHashCode(span);
            if (!_strings.TryGetValue(hash, out var value))
            {
                value = span.ToString();
                _strings[hash] = value;
            }
            return value;
        }
       
        public Order(string line)
        {
            Quantity = 0;
            Participant = "";
            Instrument = "";
           
            var input = line.AsSpan();
            var token = Token.Participant;
            int start = 0;
            int end = 0;
            for (start = 0; end < input.Length; ++end) {
                if (input[end] != ':') {
                    continue;
                }
                switch (token) {
                    case Token.Participant:
                        Participant = GetString(input.Slice(start, end - start));
                        token = Token.Instrument;
                        break;
                    case Token.Instrument:
                        Instrument = GetString(input.Slice(start, end - start));
                        token = Token.Quantity;
                        break;
                    case Token.Quantity:
                        Quantity = long.Parse(input.Slice(start, end - start));
                        token = Token.Price;
                        break;
                }
                start = end + 1;
            }

            if (token != Token.Price) {
                throw new ArgumentException($"Could not parse order '{line}'", nameof(line));
            }

            Price = double.Parse(input.Slice(start));

            RemainingQuantity = Math.Abs(Quantity);
            Generation = NextGeneration++;
        }

        public string Participant { get; } 
        public string Instrument { get; }
        public long Quantity { get; }
        public long RemainingQuantity { get; private set; }
        public double Price { get; }
        public int Generation { get; }

        public void Fill(long quantity)
        {
            RemainingQuantity -= quantity;
        }
        
        private static int NextGeneration = 1;

        public override string ToString() {
            return $"{Participant}:{Instrument}:{Quantity}:{Price}";
        }
    }    
}

File: part_2/java_solution/src/com/geh/Exchange/Order.java
package com.geh.Exchange;

import java.math.BigDecimal;
import java.util.LinkedList;

public class Order {

    private String participant;
    private String instrument;
    private long quantity;
    private long remainingQuantity;
    private double price;
    private int generation;

    private enum Token {
        Participant,
        Instrument,
        Quantity,
        Price
    };

    public void Initialise(String line) throws Exception {

        var token = Token.Participant;
        int start = 0;
        for (int i = start; (i = line.indexOf(":", start)) != -1; ) {
            switch (token) {
                case Participant:
                    this.participant = line.substring(start, i);
                    token = Token.Instrument;
                    break;
                case Instrument:
                    this.instrument = line.substring(start, i);
                    token = Token.Quantity;
                    break;
                case Quantity:
                    this.quantity = Long.parseLong(line.subSequence(start, i), 0, i - start, 10);
                    this.remainingQuantity = Math.abs(quantity);
                    token = Token.Price;
                    break;
                default:
                    break;
            }

            start = i + 1;
        }

        if (token != Token.Price) {
            throw new Exception("cannot parse: " + line);
        }

        this.price = Double.parseDouble(line.substring(start));
        this.generation = nextGeneration++;
    }

    public String getParticipant() { return participant; }
    public String getInstrument() { return instrument; }
    public long getQuantity() { return quantity; }
    public long getRemainingQuantity() { return remainingQuantity; }
    public double getPrice() { return price; }
    public int getGeneration() { return generation; }

    public void fill(long quantity)
    {
        remainingQuantity -= quantity;
    }

    private static int nextGeneration = 1;

    private static LinkedList<Order> retiredOrders = new LinkedList<>();

    public static Order get(String line) throws Exception
    {
        Order order;
        if (retiredOrders.isEmpty()) {
            order = new Order();
        }
        else {
            order = retiredOrders.pop();
        }
        order.Initialise(line);
        return order;
    }

    public static void retire(Order order)
    {
        retiredOrders.add(order);
    }

}

